File: day02/cube-conundrum.py
import sys

max_col = {
    "red": 12,
    "green": 13,
    "blue": 14
}


def check(c):
    # print (f"checking '{c}'")
    num, color = c.strip().split(' ')
    # print (f"couleur detectée: {num} cubes {color}")
    return (int(num) <= max_col[color], (int(num), color))


def check_set(s, maxset):
    cubes = s.split(',')

    setflag = True

    for c in cubes:
        flag, (num, color) = check(c)

        maxset[color] = max (
            maxset[color],
            num
        )

        setflag = setflag and flag


    return setflag


def check_game(l):
    game, reste = l.split(':')

    _,game_idstr = game.split(' ')
    ident = int(game_idstr)

    # print (f"game {ident}: {reste}")

    maxset = {
        "red": 0,
        "green": 0,
        "blue": 0
    }


    sets = reste.split(';')



    gameflag = True

    for s in sets:
        if not check_set(s, maxset):
            gameflag = False

    print ("Game valide:", ident)
    print("Mes max colors:", maxset)

    power = \
        maxset["red"] * \
        maxset["green"] * \
        maxset["blue"]

    print("Mon power:", power)

    if not gameflag:
        retval = 0
    else:
        retval = ident


    return retval, power



def main():
    somme = 0
    spower = 0
    for line in sys.stdin:
        val, power = check_game(line)
        somme += val

        spower += power

    print ("Somme totale:", somme)
    print ("Power total:", spower)








main()

File: day02/cube-conundrum.hs
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE QuasiQuotes #-}

import AoC
import Control.Applicative
import Control.Exception
import Control.Monad
import Control.Monad.ST
import qualified Data.Algebra.Boolean as B
import Data.Char
import Data.List
import Data.Maybe
import Data.Function
import Data.Array
import Data.Array.ST
import Data.Tuple.Extra
import Data.Graph

import GHC.Utils.Misc

import Debug.Trace
import qualified Data.Text as T
import qualified Data.Sequence as S
import Text.ParserCombinators.ReadP hiding (count)
import Text.Printf
-- import Text.Scanf


data Color = Red | Green | Blue deriving(Show, Eq)


strCol "green" = Green
strCol "red" = Red
strCol "blue" = Blue

parseInput = do
    games <- sepBy1 parseGame lineReturn
    skipSpaces
    eof
    return games

parseGame = do
    string "Game "
    id <- number
    colonSep
    sets <- sepBy1 parseSets semicolSep

    return (id, sets)

parseSets = sepBy1 parseColor commaSep

parseColor = do
    num <- number
    skipSpaces
    w <- word
    return (num, strCol w)

main :: IO ()
main = do
    dat <- parseContents parseInput
    print dat

    let result = part1 dat
    putStrLn $ "Résultat partie1 : " ++ show result

    let result2 = part2 dat
    putStrLn $ "Résultat partie2 : " ++ show result2


part1 games = foldl addIfValid 0 games
  where
    addIfValid acc (ident, sets) =
      acc + if validSets sets then ident else 0

    validSets sets = foldl accCheckSet True sets
    accCheckSet flag set = foldl accCheckColor flag set

    accCheckColor flag (num, col) = flag && checkCol col num

    checkCol Red n   = n <= 12
    checkCol Green n = n <= 13
    checkCol Blue n  = n <= 14



part2 games = foldl addPower 0 games
  where
    addPower pow (_, sets) = pow + power sets

    power sets = maxr * maxg * maxb
      where
        (maxr, maxg, maxb) = computeMaxCol sets
        computeMaxCol sets = 
          foldl accMaxCurrentSet  (0, 0, 0) sets

        accMaxCurrentSet themax [] = themax
        accMaxCurrentSet themax ((ncol, col):colors) =
          accMaxCurrentSet (update themax col ncol) colors

        update (mr, mg, mb) Red r = (max mr r, mg, mb)
        update (mr, mg, mb) Green g = (mr, max mg g, mb)
        update (mr, mg, mb) Blue b = (mr, mg, max mb b)


